[IV-21-22] Objetivo 6

eantoniocalo18 opened the pull request:
- closes #37 
- closes #38 

Para llevar a cabo este objetivo, se han realizado las siguientes acciones:
- Elección de Github Actions como sistema de integración continua para la ejecución de los test del código cada vez que se realice un PR.
- Elección de las diferentes versiones de Python a probar
- Elección de Circle CI para comprobar el funcionamiento del código en las diferentes versiones de python elegidas.
- Configuración de Checks API
- Actualización de la documentación del proyecto.

Lista de comprobación:
* [✓]  ¿Se han configurado varios sistemas de integración continua?
* [ ✓]  ¿Se ha configurado correctamente el *Checks API* en los sistemas en los que
   sea necesario para que aparezca correctamente en GitHub (y se pueda comprobar
   desde los tests)?
* [ ✓]  ¿Uno de los sistemas configurados permite comprobar cuales son las versiones
   del lenguaje con las que funciona correctamente nuestra aplicación?
* [ ✓]  ¿Se escogen de forma adecuada las versiones del lenguaje que se testean,
   tanto en el sistema de CI como en el contenedor Docker?

eantoniocalo18 committed:
Eliminando 'sudo docker....'

Para seguir con las buenas prácticas se ha eliminado el sudo que permitía la ejecución de docker con permisos de administrador

eantoniocalo18 committed:
Actualizando docs

Se ha actualizado la version de Python y se ha eliminado un archivo duplicado

eantoniocalo18 committed:
Eliminando archivo duplicado

eantoniocalo18 committed:
Añadiendo doc sobre integracion continua

eantoniocalo18 committed:
Creando G.Actions para test

eantoniocalo18 committed:
Actualizando info

eantoniocalo18 committed:
Creando .circleci

eantoniocalo18 committed:
Añadiendo CI a iv.yaml

eantoniocalo18 committed:
Closes #38

Se añade configuracion de poetry para que detecte invoke
Con esto, closes #38 y se avanza en M2

eantoniocalo18 committed:
Actualizando docs/CI

eantoniocalo18 committed:
Actualizando README

Se ha añadido un nuevo enlace para la nueva informacion añadida en este objetivo

eantoniocalo18 committed:
Fixing typo

JJ commented:
⛹ Revisores → @danifm1321 @lentes4k @alexespana

lentes4k reviewed (commented):
Bueno, repito como revisor tuyo jeje. En general muy buen trabajo, me gusta mucho tu manera de estructurar tu documentación y las explicaciones claras y concisas. Si bien aquí abajo te voy a dejar algunas recomendaciones o cosas que me chirrían un poco.
Por lo demás estaré pendiente de si mis compañeros ven algo que yo no he visto 👀

lentes4k commented on docs/CI.md:
```diff
@@ -0,0 +1,59 @@
+
+Este proyecto está desarrollado en Python, usando la version 3.9.
+Por lo tanto, las versiones en las que tendremos que comprobar si nuestro proyecto funciona serán las posteriores, así como la 3.8, que actualmente también es muy usada por la comunidad.
+Sin embargo, no veo la necesidad de comprobar las versiones más antiguas (como la 3.6 que se declarará obsoleta próximamente) dado que poco a poco quedarán obsoletas y no recibirán más actualizaciones, por lo que siempre tenemos que mirar hacia el futuro y asegurarnos de que las nuevas veriones de python serán compatibles con este proyecto.
```
El proceso seguido es correcto y tal, pero a la hora de documentarlo queda mas _clean_ decir que se van a testear diferentes versiones para ver en cuales funciona el código. De este modo el posible usuario sabe en que versiones del lenguaje puede instalarlo por si alguna version anterior (o posterior) tiene alguna vulnerabilidad explotable y comprobar si las dependencias son compatibles con hipoteticas versiones posteriores.

Lo documentado se me queda un poco corto en este aspecto.

lentes4k commented on .circleci/config.yml:
```diff
@@ -0,0 +1,25 @@
+      - tests:
+          matrix:
+            parameters: 
+              python-version: ["3.7", "3.8", "3.10"]
```
aunque el proyecto lo estés desarrollando tu en la versión 3.9 puede algun usuario pedirte PR y aportar su trocito usando otra versión de python. Incluso tu mismo durante la vida del proyecto puedes actualizar la versión de python de tu ordenador y desarrollarlo en otra.
Con esto quiero decir que no es mala idea la de en la matriz incluir la versión 3.9 también

eantoniocalo18 replied:
La version 3.9 es la que se usa en los test y por lo tanto no creo que sea necesario ya que comprobarías dos veces la versión 3.9

JJ replied:
> La version 3.9 es la que se usa en los test y por lo tanto no creo que sea necesario ya que comprobarías dos veces la versión 3.9

¿En qué tests?
En general, conviene que le hagáis caso a las personas que os revisan. También "es la que uso en x" no es un argumento en contra de lo que dice @lentes4k

lentes4k commented on docs/CI.md:
```diff
@@ -0,0 +1,59 @@
+    - Además, se integra con Docker y Github.
+    - Su servicio gratuito está basado en una serie de créditos iniciales que no se recargarán, por lo tanto tendríamos que pagar una vez se nos acaben.
+
+Las funcionalidades y características de Travis CI son las que necesita nuestro proyecto, sin embargo su periodo de prueba obliga a descartarlo puesto que estamos buscando una herramienta que nos permita llevar a cabo el proyecto de manera gratuita.
```
este último comentario no es del todo cierto. Sí que se puede realizar el trabajo de forma totalmente gratuita con Travis. Aunque es cierto que pide la tarjeta y algunos datos que no son agradables dar (a pesar de que los analistas de datos ya nos conocen más que nuestra familia 😂 )

eantoniocalo18 replied:
Me refería a que el periodo de prueba es tan escaso que podría quedarse corto para la realización del proyecto y tuviera que pagar

JJ replied:
Si te refieres a eso, es eso lo que tienes que poner. En general, si un revisor te pide algo debes hacer cambios en el texto para aclarar lo que te pide.

lentes4k commented on docs/CI.md:
```diff
@@ -0,0 +1,59 @@
+Ambas herramientas podrían ser utilizadas para el desarrollo del proyecto, por lo tanto, dado que no tenemos créditos ilimitados en ninguna de ellas, vamos a usar ambas.
+    - Github Actions ya ha sido utilizada anteriormente en este proyecto, por lo que ya estoy familiarizado con ella. Por lo tanto, se va a usar para la ejecución automática de los test cada vez que se realice un PR.
+
+    - Por otro lado, Circle CI será usada para la ejecución del proyecto sobre las diferentes versiones de Python.
```
quizás en algunos aspectos se hagan cortas las explicaciones. Quedan claras para nosotros que ya lo hemos realizado pero quizás alguien no demasiado puesto en el tema se pierda un poquito

eantoniocalo18 replied:
Me parece buena idea, he definido un apartado 'Definición de tareas' donde explico brévemente las tareas para las que es necesario elegir un sistema de integración continua

eantoniocalo18 committed:
Actualizando documentacion

Se ha actualizado la documentacion siguiendo las recomendaciones de @lentes4k en #39

alexespana commented on .circleci/config.yml:
```diff
@@ -0,0 +1,25 @@
+      - checkout
+      - run:
+          name: Install dependencies
+          command: pip install poetry && poetry config virtualenvs.create false && poetry install 
```
¿Has explicado por qué poner **virtualenvs.create false**? Yo sabría por qué, pero a lo mejor dentro de unos meses me quedo pensando por qué esa elección :thinking: .

danifm1321 replied:
Deberías poner por qué has puesto ese virtualenvs.create false, y si hay alguna alternativa a ponerlo.

alexespana commented on docs/CI.md:
```diff
@@ -0,0 +1,67 @@
+## Definición de requisitos
+
+Los requisitos que nos servirán para valorar positiva o negativamente una herramienta de integración continua serán los siguientes:
+- Lenguajes soportados (versiones de Python)
```
¿Hay alguno que no lo soporte?

alexespana commented on docs/CI.md:
```diff
@@ -0,0 +1,67 @@
+
+# Versiones del lenguaje
+
+Este proyecto está desarrollado en Python, usando la version 3.9.
```
Aquí podrías incluir una referencia a las dependencias que indica poetry.

alexespana commented on docs/CI.md:
```diff
@@ -0,0 +1,67 @@
+# Versiones del lenguaje
+
+Este proyecto está desarrollado en Python, usando la version 3.9.
+Por lo tanto, se ve la necesidad de testear el código del proyecto en las versiones anteriores y posteriores a la utilizada para comprobar su correcto funcionamiento.
```
Realmente testeas la aplicación con distintas versiones del lenguaje para ver con cuál de ellas funciona tu aplicación, no es simplemente ver cuál usas y testear versiones inferiores y posteriores.

JJ replied:
> Objetivo listo para revisión. @JJ

¿Seguro? ¿No querrás hacer algúna revisión antes, como esta y las de más arriba?

alexespana reviewed (commented):
Buenas @eantoniocalo18 , en general lo veo aceptable, parece que entiendes los conceptos básicos que transmite el objetivo. En la documentación añadiría algún enlace en el que te hayas apoyado para realizar las configuraciones con CircleCI y GA. Otra cosa que me gustaría añadir, el milestone al que está asignado este PR no parece incluir el desarrollo realizado, yo lo cambiaría. Échale un ojo a los comentarios que te hago por aquí abajo :arrow_down: .

eantoniocalo18 commented:
Objetivo listo para revisión. @JJ

eantoniocalo18 committed:
Actualizando doc

Se ha actualizado la documentacion en base a los comentarios realizados por @alexespana en #39

danifm1321 commented on docs/CI.md:
```diff
@@ -0,0 +1,68 @@
+# Versiones del lenguaje
+
+Este proyecto está desarrollado en Python, usando la version 3.9.
+Por lo tanto, se ve la necesidad de testear el código del proyecto en las versiones anteriores y posteriores a la utilizada para comprobar su correcto funcionamiento.
```
Yo le daría una pequeña vuelta más a la justificación de las versiones elegidas. Como me dijo a mi JJ en mi PR, cada test cuesta dinero, hay que tener un buen criterio para elegir las versiones que testean.

danifm1321 reviewed (changes requested):
Parece que llego un poco tarde, y mis compañeros ya han hecho un buen trabajo con las revisiones. Sin embargo, quiero hacer énfasis en dos cosas que te han comentado que me parecen importantes.

eantoniocalo18 committed:
Update CI.md, fixing typo

eantoniocalo18 committed:
Actualizando doc y corrigiendo errores

Se ha actualizado la documentacion en base a los comentarios de los revisores y se han corregido los errores correspondientes

eantoniocalo18 commented:
Cambios en la documentación realizados.
Se han corregido los errores que me habéis comentado y por otra parte he creado un nuevo archivo de documentación Dependencias.md para explicar como se instalan y también el porqué del uso de virtual.envs.
¿Podríais echarle un último vistazo antes de que se lo envíe al profesor? @lentes4k @alexespana @danifm1321

alexespana commented on docs/dependencias.md:
```diff
@@ -0,0 +1,20 @@
+
+Con esto, ya tendríamos las dependencias instaladas. Sin embargo, estas dependencias estarían instaladas en un entorno virtual propio que crea poetry.
+Esto nos supone un problema, puesto que las pruebas que se van a realizar al código van a ser en un entorno virtualizado (un dokcer) y por otra parte la instalación con Poetry nos creará un nuevo entorno virtualizado. 
+Es por ello, que tenemos que pasarle el flag 'virtual.envs=False', para que Poetry no instale las dependencias en su propio entorno virtual sino en el entorno que nosotros estamos trabajando.
```
Esto está mal, el flag es **virtualenvs.create** y el valor será false en todo caso.

eantoniocalo18 replied:
Cierto, tienes razon :+1:

eantoniocalo18 committed:
Creando dependencias.md

Se ha visto la necesidad de documentar el proceso de instalacion de dependencias, asi como una breve explicacion de como trabaja el gestor de dependencias elegido

alexespana reviewed (changes requested):
El nuevo archivo (**dependencias.md**) lo borraría a mi parecer, ya que es más bien parte de los objetivos anteriores y es muy escueto lo que se comenta. La parte más importante del archivo (la que se te comentó que añadieses) es el último párrafo, simplemente podrías añadir una pequeña **nota** o un **nuevo apartado** al final del archivo **CI.md** donde ponerlo.

Por otro lado, el milestone que describe el desarrollo del objetivo no parece que lo contenga ni de forma explícita ni implícita, cámbialo por favor (crea uno que lo contenga). Ánimo, casi está :muscle: .

eantoniocalo18 commented:
Cambios sugeridos por @alexespana realizados.
Esperando el visto bueno de @lentes4k y  @danifm1321

lentes4k reviewed (approved):
Yo lo veo ya bastante bien. Se han resuelto los comentarios de los compañeros, así que 👍 😎

eantoniocalo18 committed:
Actualizando doc

Se ha eliminado el archivo de dependencias, incluyendo la parte mas relevante en el archivo CI.md

danifm1321 reviewed (approved):
Bastante bien, todo resuelto

eantoniocalo18 commented:
Todos los cambios realizados.
@JJ Listo para revisión.

JJ reviewed (approved):
Aparte de los comentarios que te he dejado, que no evitar pasar el objetivo, estás testeando demasiadas versiones de Python, sobre todo porque puede que la 3.9 (o la que uses en el Dockerfile) quizás se está probando dos veces. Por lo demás, buen trabajo tuyo y de los revisores.
